File: lib/noizu_prompt_lingua/npl/resolver.ex
defmodule NoizuPromptLingua.NPL.Resolver do
  @moduledoc false

  alias NoizuPromptLingua.NPL.Parser.{Component, Expression}

  require Logger

  @section_files %{
    "syntax" => "syntax.yaml",
    "declarations" => "declarations.yaml",
    "directives" => "directives.yaml",
    "prefixes" => "prefixes.yaml",
    "prompt-sections" => "prompt-sections.yaml",
    "special-sections" => "special-sections.yaml",
    "pumps" => "pumps.yaml",
    "fences" => "fences.yaml"
  }

  defmodule ResolvedComponent do
    @moduledoc false
    defstruct [
      :section,
      :name,
      :slug,
      :brief,
      :description,
      syntax: [],
      examples: [],
      labels: [],
      require: [],
      priority_filtered: false
    ]

    @type t :: %__MODULE__{
            section: String.t(),
            name: String.t(),
            slug: String.t(),
            brief: String.t(),
            description: String.t(),
            syntax: [map()],
            examples: [map()],
            labels: [String.t()],
            require: [String.t()],
            priority_filtered: boolean()
          }
  end

  defstruct [:npl_dir, cache: %{}]

  @type t :: %__MODULE__{npl_dir: String.t(), cache: map()}

  @spec new(String.t()) :: t()
  def new(npl_dir), do: %__MODULE__{npl_dir: npl_dir}

  @spec resolve(t(), Expression.t()) :: {:ok, [ResolvedComponent.t()]} | {:error, String.t()}
  def resolve(resolver, %Expression{} = expression) do
    with {:ok, resolver, components_map} <- process_additions(resolver, expression.additions),
         {:ok, resolver, components_map} <- process_subtractions(resolver, expression.subtractions, components_map) do
      result = build_ordered_result(resolver, expression.additions, components_map)
      {:ok, result}
    end
  end

  defp process_additions(resolver, additions) do
    Enum.reduce_while(additions, {:ok, resolver, %{}}, fn addition, {:ok, res, map} ->
      case resolve_single(res, addition) do
        {:ok, res, resolved} ->
          new_map =
            Enum.reduce(resolved, map, fn comp, acc ->
              Map.put(acc, {comp.section, comp.slug}, comp)
            end)

          {:cont, {:ok, res, new_map}}

        {:error, _} = err ->
          {:halt, err}
      end
    end)
  end

  defp process_subtractions(resolver, subtractions, components_map) do
    new_map =
      Enum.reduce(subtractions, components_map, fn sub, map ->
        if sub.component == nil do
          Map.reject(map, fn {{section, _slug}, _comp} -> section == sub.section end)
        else
          key = {sub.section, sub.component}

          if Map.has_key?(map, key) do
            Map.delete(map, key)
          else
            Logger.warning(
              "Cannot subtract '#{sub.component}' from '#{sub.section}' - not in loaded set. Ignoring."
            )

            map
          end
        end
      end)

    {:ok, resolver, new_map}
  end

  defp build_ordered_result(resolver, additions, components_map) do
    {result, _seen} =
      Enum.reduce(additions, {[], MapSet.new()}, fn addition, {acc, seen} ->
        if addition.component == nil do
          case load_section(resolver, addition.section) do
            {:ok, _resolver, data} ->
              components_data = Map.get(data, "components", [])

              Enum.reduce(components_data, {acc, seen}, fn comp_data, {inner_acc, inner_seen} ->
                slug = Map.get(comp_data, "slug", Map.get(comp_data, "name", ""))
                key = {addition.section, slug}

                if Map.has_key?(components_map, key) and not MapSet.member?(inner_seen, key) do
                  {[components_map[key] | inner_acc], MapSet.put(inner_seen, key)}
                else
                  {inner_acc, inner_seen}
                end
              end)

            {:error, _} ->
              {acc, seen}
          end
        else
          key = {addition.section, addition.component}

          if Map.has_key?(components_map, key) and not MapSet.member?(seen, key) do
            {[components_map[key] | acc], MapSet.put(seen, key)}
          else
            {acc, seen}
          end
        end
      end)

    Enum.reverse(result)
  end

  defp resolve_single(resolver, %Component{} = component) do
    with {:ok, resolver, data} <- load_section(resolver, component.section) do
      components_data = Map.get(data, "components", [])

      if component.component == nil do
        resolved =
          Enum.map(components_data, fn comp_data ->
            resolve_component(component.section, comp_data, component.priority_max)
          end)

        {:ok, resolver, resolved}
      else
        slug_match = fn c ->
          Map.get(c, "slug", Map.get(c, "name", "")) == component.component
        end

        case Enum.find(components_data, slug_match) do
          nil ->
            available =
              components_data
              |> Enum.map(&Map.get(&1, "slug", Map.get(&1, "name", "")))
              |> Enum.filter(&(&1 != ""))
              |> Enum.join(", ")

            {:error,
             "Component '#{component.component}' not found in section '#{component.section}'. Available components: #{available}"}

          comp_data ->
            {:ok, resolver, [resolve_component(component.section, comp_data, component.priority_max)]}
        end
      end
    end
  end

  defp resolve_component(section, comp_data, priority_max) do
    examples = Map.get(comp_data, "examples", [])

    {filtered_examples, priority_filtered} =
      case priority_max do
        nil ->
          {examples, false}

        max when max < 0 ->
          {[], true}

        max ->
          filtered = Enum.filter(examples, fn ex -> Map.get(ex, "priority", 0) <= max end)
          {filtered, true}
      end

    %ResolvedComponent{
      section: section,
      name: Map.get(comp_data, "name", ""),
      slug: Map.get(comp_data, "slug", Map.get(comp_data, "name", "")),
      brief: Map.get(comp_data, "brief", ""),
      description: Map.get(comp_data, "description", ""),
      syntax: Map.get(comp_data, "syntax", []),
      examples: filtered_examples,
      labels: Map.get(comp_data, "labels", []),
      require: Map.get(comp_data, "require", []),
      priority_filtered: priority_filtered
    }
  end

  defp load_section(%__MODULE__{cache: cache} = resolver, section) do
    case Map.fetch(cache, section) do
      {:ok, data} ->
        {:ok, resolver, data}

      :error ->
        filename = Map.get(@section_files, section)

        if filename == nil do
          {:error, "Unknown section: '#{section}'"}
        else
          filepath = Path.join(resolver.npl_dir, filename)

          if File.exists?(filepath) do
            case YamlElixir.read_from_file(filepath) do
              {:ok, data} ->
                new_resolver = %{resolver | cache: Map.put(cache, section, data)}
                {:ok, new_resolver, data}

              {:error, reason} ->
                {:error, "Invalid YAML in #{filename}: #{inspect(reason)}"}
            end
          else
            {:error, "Section file not found: #{filename}. Expected at: #{filepath}"}
          end
        end
    end
  end
end

File: projects/NoizuPromptLingo/src/npl_mcp/npl/resolver.py
"""
NPL Section Resolver.

Resolves NPL expressions to component data by loading YAML files
and applying filters.
"""

from dataclasses import dataclass
from pathlib import Path
from typing import Dict, Any, List, Optional
import yaml
import logging

from .parser import NPLSection, NPLExpression, NPLComponent
from .filters import filter_by_priority
from .exceptions import NPLResolveError

logger = logging.getLogger(__name__)


@dataclass
class ResolvedComponent:
    """A resolved NPL component with its data."""
    section: NPLSection
    name: str
    slug: str
    brief: str
    description: str
    syntax: List[Dict[str, Any]]
    examples: List[Dict[str, Any]]
    labels: List[str]
    require: List[str]
    priority_filtered: bool  # True if examples were filtered by priority


# Mapping from section enum to YAML filename
_SECTION_FILES = {
    NPLSection.SYNTAX: "syntax.yaml",
    NPLSection.DECLARATIONS: "declarations.yaml",
    NPLSection.DIRECTIVES: "directives.yaml",
    NPLSection.PREFIXES: "prefixes.yaml",
    NPLSection.PROMPT_SECTIONS: "prompt-sections.yaml",
    NPLSection.SPECIAL_SECTIONS: "special-sections.yaml",
    NPLSection.PUMPS: "pumps.yaml",
    NPLSection.FENCES: "fences.yaml",
}


class NPLResolver:
    """Resolves NPL expressions to component data."""

    def __init__(self, npl_dir: Path):
        """Initialize resolver with path to NPL YAML files.

        Args:
            npl_dir: Path to directory containing NPL YAML files
        """
        self.npl_dir = Path(npl_dir)
        self._cache: Dict[NPLSection, Dict[str, Any]] = {}

    def _load_section(self, section: NPLSection) -> Dict[str, Any]:
        """Load and cache a section's YAML data.

        Args:
            section: The section to load

        Returns:
            Parsed YAML data

        Raises:
            NPLResolveError: If file not found or invalid YAML
        """
        # Return cached data if available
        if section in self._cache:
            return self._cache[section]

        filename = _SECTION_FILES[section]
        filepath = self.npl_dir / filename

        if not filepath.exists():
            raise NPLResolveError(
                f"Section file not found: {filename}. "
                f"Expected at: {filepath}"
            )

        try:
            with open(filepath, 'r', encoding='utf-8') as f:
                data = yaml.safe_load(f)
        except yaml.YAMLError as e:
            raise NPLResolveError(
                f"Invalid YAML in {filename}: {str(e)}"
            )

        # Cache the loaded data
        self._cache[section] = data
        return data

    def get_section_components(self, section: NPLSection) -> List[str]:
        """Get list of component slugs in a section.

        Args:
            section: The section to query

        Returns:
            List of component slugs
        """
        data = self._load_section(section)
        components = data.get('components', [])
        return [c.get('slug', '') for c in components if c.get('slug')]

    def validate_component(self, section: NPLSection, component: str) -> bool:
        """Check if component exists in section.

        Args:
            section: The section to check
            component: The component slug to validate

        Returns:
            True if component exists, False otherwise
        """
        slugs = self.get_section_components(section)
        return component in slugs

    def _resolve_component(
        self,
        section: NPLSection,
        component_data: Dict[str, Any],
        priority_max: Optional[int]
    ) -> ResolvedComponent:
        """Convert raw component data to ResolvedComponent.

        Args:
            section: The section this component belongs to
            component_data: Raw component data from YAML
            priority_max: Maximum priority for filtering examples (None = no filter)

        Returns:
            ResolvedComponent with filtered examples
        """
        examples = component_data.get('examples', [])
        priority_filtered = False

        # Apply priority filter if specified
        if priority_max is not None:
            examples = filter_by_priority(examples, priority_max)
            priority_filtered = True

        return ResolvedComponent(
            section=section,
            name=component_data.get('name', ''),
            slug=component_data.get('slug', ''),
            brief=component_data.get('brief', ''),
            description=component_data.get('description', ''),
            syntax=component_data.get('syntax', []),
            examples=examples,
            labels=component_data.get('labels', []),
            require=component_data.get('require', []),
            priority_filtered=priority_filtered,
        )

    def _resolve_single(self, component: NPLComponent) -> List[ResolvedComponent]:
        """Resolve a single NPLComponent to ResolvedComponent(s).

        Args:
            component: The component to resolve

        Returns:
            List of ResolvedComponent (one if specific component, many if entire section)

        Raises:
            NPLResolveError: If component not found
        """
        data = self._load_section(component.section)
        components_data = data.get('components', [])

        if component.component is None:
            # Load entire section
            return [
                self._resolve_component(
                    component.section,
                    comp_data,
                    component.priority_max
                )
                for comp_data in components_data
            ]
        else:
            # Load specific component
            for comp_data in components_data:
                if comp_data.get('slug') == component.component:
                    return [
                        self._resolve_component(
                            component.section,
                            comp_data,
                            component.priority_max
                        )
                    ]

            # Component not found
            available = self.get_section_components(component.section)
            raise NPLResolveError(
                f"Component '{component.component}' not found in section '{component.section.value}'. "
                f"Available components: {', '.join(available)}"
            )

    def resolve(self, expression: NPLExpression) -> List[ResolvedComponent]:
        """Resolve expression to list of components.

        Applies additions first, then subtractions.
        Validates all component references exist.

        Args:
            expression: The parsed expression to resolve

        Returns:
            List of resolved components

        Raises:
            NPLResolveError: If component not found
        """
        # Track components by (section, slug) to handle additions/subtractions
        components_map: Dict[tuple, ResolvedComponent] = {}

        # Process additions
        for addition in expression.additions:
            resolved = self._resolve_single(addition)
            for comp in resolved:
                key = (comp.section, comp.slug)
                # Later additions can override earlier ones (for re-adding)
                components_map[key] = comp

        # Process subtractions
        for subtraction in expression.subtractions:
            if subtraction.component is None:
                # Subtract entire section
                keys_to_remove = [
                    key for key in components_map
                    if key[0] == subtraction.section
                ]
                for key in keys_to_remove:
                    del components_map[key]
            else:
                # Subtract specific component
                key = (subtraction.section, subtraction.component)
                if key in components_map:
                    del components_map[key]
                else:
                    # Subtracting something not loaded - warn but don't error
                    logger.warning(
                        f"Cannot subtract '{subtraction.component}' from '{subtraction.section.value}' "
                        "- not in loaded set. Ignoring."
                    )

        # Return components in a stable order
        # Preserve the order from additions (first occurrence)
        result = []
        seen = set()
        for addition in expression.additions:
            if addition.component is None:
                # Entire section - add all in YAML order
                try:
                    data = self._load_section(addition.section)
                    for comp_data in data.get('components', []):
                        slug = comp_data.get('slug', '')
                        key = (addition.section, slug)
                        if key in components_map and key not in seen:
                            result.append(components_map[key])
                            seen.add(key)
                except NPLResolveError:
                    pass  # Section might have been fully subtracted
            else:
                # Specific component
                key = (addition.section, addition.component)
                if key in components_map and key not in seen:
                    result.append(components_map[key])
                    seen.add(key)

        return result
